### PR TITLE
Add search results pagination

### PR DESCRIPTION
**Pull request by nicastelo**
#### Issue reference:

* [#19](https://github.com/WorldBrain/webmarks/issues/19)

---

#### Description:

* Add search results pagination.

---

#### Reviewers:

* @oliversauter
* @dev-tim

---

#### Notes:

* Added more dummy results on the server.

---

#### Tasks:

  - [x] Create prev & next buttons .

  - [x] Create page's links.

  - [x] Load results per page.

---

#### Risk:

* Low

### CODE REVIEW

**Comment by oliversauter** [comment]
@nicastelo 
Great thanks for your first PR :) Wohooo! Now you are officially on board.

@tobisanya @hoshomoh 
This is the PR related to your work this week with https://github.com/WorldBrain/cortex/issues/37

@dev-tim 
do you think you have the time to review this PR this week?

**Comment by dev-tim** [comment]
Please disable eslint. 
In overall LGTM, we could use https://github.com/AdeleD/react-paginate. But if custom impl works - :shipit:

File: src/components/search-result-item/search-result-item.js [hunk 1]
```diff
@@ -5,17 +5,20 @@ class SearchResultItem extends React.Component {
...
     url: React.PropTypes.string.isRequired,
-    description: React.PropTypes.string.isRequired
+    description: React.PropTypes.string.isRequired,
   };
 
   static defaultProps = {
     title: '"Generation 25": Gelebte Einheit - Gelernte Grenzen',
     url: 'http://www.dw.com/de/s%C3%BCndenbock-der-eu-junckers-dilemma/a-19391383',
-    description: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore etdolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip exea commodo consequat."
+    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, ' +
```
Comment (obsidianart) [inline src/components/search-result-item/search-result-item.js]: I would generally avoid this approach, seems the IDE was complaining more than a needed change. consider also ` if you want multiline.

File: src/components/search-result-list/search-result-list.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import SearchResultItem from '../search-result-item';
+import SearchResultPagination from '../search-result-pagination';
+
+function render(data) {
+  const list = data.results.map((result, i) => {
+    let title;
+    let url;
+    let description;
+
+    if (typeof result !== 'undefined') {
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: if this is possible probably better to filter first an loop after
```
data.results.filter(result=>!!result).map(....
```
or if you are worried precisely of undefined ( I would say any falsy should be discarded no matter what)
```
data.results.filter(result=>!!typeof result !== 'undefined').map(....
```

Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: Unless you actually want empty searchResultItem, in which case I'm

File: src/components/search-result-list/search-result-list.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        />
+      </div>
+    </div>
+  );
+}
+
+function SearchResultList(props) {
+  const results = props.results.json;
+
+  if (!Array.isArray(results)) {
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: What can it be? Why do we need this safeguard? do we want to at least ```console.warn``` something went wrong? My main worry is that we discard errors which shouldn't happen without noticing for a long time and making debugging of a problem way longer.
Also, if props.results is undefined we crash on line 44, so this condition only saves for existing results but not in array form

File: src/components/search-result-list/search-result-list.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    return null;
+  }
+
+  const actualPage = Math.round(parseInt(props.page, 10));
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: This is a code smell and I'm sorry if I'm wrong, but I fell like this should either be Math.floor or Math.ceiling. I might be completely wrong

File: src/components/search-result-list/search-result-list.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+  const resultsPerPage = 5;
+  const pagesQty = Math.round(results.length / resultsPerPage);
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: This is a code smell and I'm sorry if I'm wrong, but I fell like this should either be Math.floor or Math.ceiling. I might be completely wrong

File: src/components/search-result-list/search-result-list.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  const pagesPerPage = pagesQty > 10 ? 10 : pagesQty;
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: What about ```Math.min(pagesQty,10)```?

File: src/components/search-result-list/search-result-list.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  const lastPage = pagesQty;
+
+  if (actualPage < 1 || actualPage > lastPage) {
+    props.paginate();
+  }
+
+  const start = actualPage === 1 ? 0 : resultsPerPage * (actualPage - 1);
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: isn't ```resultsPerPage * (actualPage - 1)``` 0 when actualPage is 1? why the ternary?

File: src/components/search-result-list/search-result-list.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  const end = start + resultsPerPage;
+  const pageResults = results.slice(start, end);
+
+  const pages = [...Array(pagesQty)].map((_, i) => {
+    const pageNumber = i + 1;
+    let disable = false;
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: ```disable = actualPage === pageNumber``` ?

File: src/components/search-result-list/search-result-list.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    return {
+      page: pageNumber,
+      disable,
+    };
+  });
+
+  let actualPages;
+  if ((lastPage - actualPage) < Math.round(pagesPerPage / 2)) {
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: as above, Math.round smells incorrect

File: src/components/search-result-list/search-result-list.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    const actualFirstPage = lastPage - pagesPerPage;
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: can't declare variable inside if condition, or constant as far as I understand ecma (This is actually a gray area and works in unexpected way, different engine create the context differently)

File: src/components/search-result-list/search-result-list.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+  if (actualPage < 1 || actualPage > lastPage) {
+    props.paginate();
+  }
+
+  const start = actualPage === 1 ? 0 : resultsPerPage * (actualPage - 1);
+  const end = start + resultsPerPage;
+  const pageResults = results.slice(start, end);
+
+  const pages = [...Array(pagesQty)].map((_, i) => {
```
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]:
```
const pages = [...Array(pagesQty)].map((_, i) => ({
    page: i+1,
    disable = actualPage === pageNumber
})
```
?

File: src/components/search-result-pagination/search-result-pagination.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
+import React from 'react';
+
+function pages(data, paginate) {
+  return data.map((page, i) => (
+    <li className="search-pagination-page" key={i}>
```
Comment (obsidianart) [inline src/components/search-result-pagination/search-result-pagination.js]: key should be more than i, i can easily be used by another loop. maybe ```key={`search-page${i}````

File: src/components/search-result-pagination/search-result-pagination.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        onClick={(e) => {
+          e.preventDefault();
+          paginate(page);
+        }}
+      >Prev</a>
+    </li>
+  );
+}
+
+function next(actualPage, lastPage, paginate) {
```
Comment (obsidianart) [inline src/components/search-result-pagination/search-result-pagination.js]: The DRY seems violated, those 3 functions (pages map, prev, next) produce 90% the same code. is it possible to merge them?

File: src/containers/main/home/home-section.js [hunk 14]
```diff
@@ -1,23 +1,24 @@
 import React from 'react';
-import SearchBar  from '../../../components/search-bar';
-import {querySearch} from '../../../redux/actions/search';
+import SearchBar from '../../../components/search-bar';
+import { querySearch } from '../../../redux/actions/search';
 import { connect } from 'react-redux';
 
 const mapStateToProps = (state) => ({
-  searchResults: state.searchResults
+  searchResults: state.searchResults,
```
Comment (obsidianart) [inline src/containers/main/home/home-section.js]: why this change?

File: src/containers/main/home/home-section.js [hunk 15]
```diff
@@ -33,7 +34,7 @@ class HomeSection extends React.Component {
             </div>
             <div className="home-search__search-container">
               <div className="home-search__search-input">
-                <SearchBar onSearch={this.onSearchQuery.bind(this)}/>
+                <SearchBar onSearch={() => { this.onSearchQuery(); }} />
```
Comment (obsidianart) [inline src/containers/main/home/home-section.js]: why this change? the bind pattern is really common in react. Does this project decided against it?

File: src/containers/main/search-results/search-results-section.js [hunk 16]
```diff
@@ -37,11 +55,11 @@ class SearchResultsSection extends React.Component {
       <div className="search-results">
         <div className="search-results__top">
           <div className="search-logo">
-            <img src={logo} className="search-logo__img" alt="webmarks"/>
+            <img src={logo} className="search-logo__img" alt="webmarks" />
           </div>
           <div className="search-container">
             <div className="search-bar">
-              <SearchBar value={query} onSearch={this.onSearchQuery.bind(this)}/>
+              <SearchBar value={query} onSearch={() => { this.onSearchQuery(); }} />
```
Comment (obsidianart) [inline src/containers/main/search-results/search-results-section.js]: as above, what against the bind pattern?

**Comment by oliversauter** [comment]
wow thank @obsidianart for your extensive code review!

**Comment by obsidianart** [comment]
first level code review, style and functionality, second step is structure after the first pass is done and it's easier to read. @nicastelo sorry if it seems pedantic, I did a similar level to what I do at work (slightly less strict)

**Comment by nicastelo** [comment]
@obsidianart Wow nice review, thanks man.

File: src/containers/main/search-results/search-results-section.js (hunk 16 above)
Comment (nicastelo) [inline src/containers/main/search-results/search-results-section.js]: These changes are because of eslint rules. The project is using the aribnb's eslint rules.

Travis is failing because of that.

I think we have 3 options. Follow these rules, define our own rules or take off eslint.

File: src/containers/main/home/home-section.js (hunk 15 above)
Comment (nicastelo) [inline src/containers/main/home/home-section.js]: Because of eslint https://github.com/WorldBrain/webmarks/pull/34#discussion_r77932603

File: src/containers/main/home/home-section.js (hunk 14 above)
Comment (nicastelo) [inline src/containers/main/home/home-section.js]: Beacuse of eslint https://github.com/WorldBrain/webmarks/pull/34#discussion_r77932603

File: src/components/search-result-list/search-result-list.js (hunk 11 above)
Comment (nicastelo) [inline src/components/search-result-list/search-result-list.js]: I use [...Array(pagesQty)] to create an iteration over pagesQty in order to make the pagination links.

File: src/components/search-result-list/search-result-list.js (hunk 10 above)
Comment (nicastelo) [inline src/components/search-result-list/search-result-list.js]: As a far as I know const and let are block-scoped variables

File: src/components/search-result-list/search-result-list.js (hunk 2 above)
Comment (nicastelo) [inline src/components/search-result-list/search-result-list.js]: Why iterate two times over the same thing?

correct me if I'm wrong but I think Array.filter is an iteration like Array.map

File: src/components/search-result-item/search-result-item.js (hunk 1 above)
Comment (nicastelo) [inline src/components/search-result-item/search-result-item.js]: Because of eslint https://github.com/WorldBrain/webmarks/pull/34#discussion_r77932603

File: src/components/search-result-list/search-result-list.js (hunk 3 above)
Comment (nicastelo) [inline src/components/search-result-list/search-result-list.js]: This is because props.results.json can be 'undefined'.

In that case, without this safeguard we crash on line 6

Comment (nicastelo) [inline src/components/search-result-list/search-result-list.js]: This is because props.results.json can be 'undefined'
In that case, without this safeguard we crash on line 6.

File: src/components/search-result-item/search-result-item.js (hunk 1 above)
Comment (obsidianart) [inline src/components/search-result-item/search-result-item.js]: if linter can be discussed I would disable it, at least on the 80 characters rules and other ideas which make little sense with modern monitor and IDE (semicolon, inline functions, etc). If linter is kept, probably a pre-commit hook is good. modifying for linting create a future problem, git annotate will show your name next to those line and make it hard to track the original owner.

File: src/components/search-result-list/search-result-list.js (hunk 2 above)
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: write code for humans, not machines.

We are discussing optimisation of a 10 element loop, agains readability. 

I avoid this kind or optimiziation and, to be fair, this if will generate an empty element which will generate an empty searchResultItem which will probably crate an empty element or trigger other checks which will, by themselves, be way slower than this optimization.

filter, loops, etc, are all a form of for but might be optimized by V8 (filter, for example, can be parallelized)

If a difference exist, we are talking about ms, I advice this article http://mrale.ph/blog/2012/12/15/microbenchmarks-fairy-tale.html

File: src/components/search-result-list/search-result-list.js (hunk 3 above)
Comment (obsidianart) [inline src/components/search-result-list/search-result-list.js]: how can it be undefined? this should be a requirement for the component and any check on needed props should happen in one place.
Think about how props work in React, you specify what you want in them in details in propTypes, so if someone wants to use you it should specify them. Do you have an actual use for this componenent if results is undefined or it is a spurious state?

File: src/containers/main/search-results/search-results-section.js (hunk 16 above)
Comment (obsidianart) [inline src/containers/main/search-results/search-results-section.js]: sorry I had no idea, I assumed the current code base was stable. I commented about that once. thanks

**Comment by dev-tim** [comment]
@obsidianart  First of all nice review! But at this stage I'd not be so strict, it's an MVP simple front-end to test the idea if it works at all.  At this stage I'd put more attention tests and actual data + actual features. 

From developers point of view you are right, from pragmatic point I'd use library, but as soon as it works, feel free to shoot. 

@nicastelo Could you disable linter check on travis? Seems like we don't obey to the styleguide atm, so it's just producing noice.

**Comment by obsidianart** [comment]
@dev-tim It wasn't too strict, I'm much stricter at work, this is for me to learn the decision process etc and for nicastelo to be able to discuss. Changing code later is hard anyway. Feel free to merge if you want.

**Comment by oliversauter** [comment]
@obsidianart 
@dev-tim 
I am actually also pro the rather stricter review, even if it is just for a prototype. This will be a win win for all of the people. As @obsidianart said, @nicastelo will be able to learn from it and discuss. And for the project we are rather a little bit slower but have more solid code. 
Even if it is just a MVP, we are not wanting to throw away everything afterwards. 

Thanks for your review all of you!

**Comment by oliversauter** [comment]
@nicastelo do you feel comfortable to merge this / What else is to do?

**Comment by obsidianart** [comment]
Map should be change or my comment replied. I sent @nicastelo a js example yesterday to show better what I meant, after that I'm fine

**Comment by nicastelo** [comment]
@oliversauter @obsidianart Sorry for the delay, I was very busy with other things.

**Comment by obsidianart** [comment]
good to go for me

**Comment by oliversauter** [comment]
Merging now. Thanks @nicastelo for your contribution and @obsidianart, @dev-tim for your thorough review!
